File: codingame/java/src/main/HackingRobberCity.java
/**
 * At RobberCity, it is really hard to send a parcel safely. Postmen are thieves, and you can be sure that without a proper lock, your box will be opened and emptied prior to "delivery".
 * This is a famous riddle: how can Alice send a parcel to Bob, without having to send the lock key?
 *
 * The answer is:
 * * step 1: Alice locks the box using her own lock and send it to Bob.
 * * step 2: Bob receives the unopened parcel, but cannot open it without the key. He add his own lock on the parcel and send it back to Alice.
 * * step 3: Alice gets her parcel back with the 2 locks. She can now remove hers and send it Bob again.
 * * step 4: Bob receives the parcel with his lock, that he can open.
 *
 * Now, the electronic version!
 * This time, Alice wants to send secured data to Bob. They both read this wikipedia article:
 * https://en.wikipedia.org/wiki/XOR_cipher#Use_and_security about the XOR cipher
 * With a key that is truly random, the result is a one-time pad, which is unbreakable in theory.
 * Wow. Unbreakable.
 *
 * Alice and Bob decided to use the riddle solution applied to XOR.
 * * step 1: Alice encodes her message with her random key, as long as the message itself. She sends it to Bob.
 * * step 2: Bob encodes the ciphered message with his own random key, as long as the message itself. He sends it back to Alice.
 * * step 3: Alice decodes the message with her initial key, and sends it to Bob.
 * * step 4: Bob decodes the message with his initial key, and gets the clear text.
 *
 * And it works! XOR is both commutative and associative, and A XOR 0 = A and A XOR A = 0. Hence
 * Message XOR AliceKey XOR BobKey XOR AliceKey XOR BobKey
 *  = Message XOR (AliceKey XOR AliceKey) XOR (BobKey XOR BobKey)
 *  = Message XOR 0 XOR 0
 *  = Message
 *
 * You are a man-in-the-middle, intercepting any message between Alice and Bob.
 * (Btw, your existence proves that they both really had good reasons to put a data cipher process in place...)
 * So, you intercept the 3 messages message1, message2 and message3.
 *
 * Your goal is to be smarter than the smarties, and to break their code.
 *
 * Note: intercepted messages are made of bytes (values from 0 to 255), that will be displayed in hexadecimal form.
 * Thus, each message will be an hexadecimal string of size 2n, each digit pair corresponds to a byte (from 00=0 to FF=255).
 *
 * Once you get the clear message bytes, convert to clear text using ASCII encoding.
 * Input
 * Line 1: message1, the first message Alice sends to Bob (hexadecimal)
 * Line 2: message2, the second message Bob sends back (hexadecimal)
 * Line 3: message3, the last message Alice sends (hexadecimal)
 * Output
 * The clear text message
 * Constraints
 * Message length ≤ 250 chars (i.e. message1, message2 and message3 lengths are ≤ 500 hexadecimal digits)
 * */

import java.util.stream.Collectors;
import java.util.stream.IntStream;

class HackingRobberCity {

    public String process1(String msg1, String msg2, String msg3) {
        // process
        String tmp = xorTwoHexadecimalStr(msg1, msg2); // tmp = msg1 xor msg2 = B
        String resHexStr = xorTwoHexadecimalStr(tmp, msg3); // res = tmp xor msg3 = C
        // output
        return hexDecimal2CharStr(resHexStr);
    }

    private String xorTwoHexadecimalStr(String str1, String str2){
        byte[] byteArr1 = hexStr2ByteArray(str1);
        byte[] byteArr2 = hexStr2ByteArray(str2);
        // XOR between two byte[]
        return IntStream.range(0, byteArr1.length)
                .mapToObj(i -> String.format("%02x", (byte)(byteArr1[i] ^ byteArr2[i]))) // 0xFF -> "FF"
                .collect(Collectors.joining());
    }

    // hexadecimal with base-16: 00=0, FF=255 which uses 0-9 and A-F/a-f to represent value 0-15
    private byte[] hexStr2ByteArray(String hexStr) {
        int len = hexStr.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len-1; i += 2) {
            // e.g "FF" [as 0xFF] -> 15 * 2^4 + 15 = int(255) -> (byte)255 = 0xFF
            data[i/2] = (byte) (
                (Character.digit(hexStr.charAt(i), 16) << 4)
                    + Character.digit(hexStr.charAt(i + 1), 16));
        }
        return data;
    }

    // hexadecimal with base-16: 00=0, FF=255
    private String hexDecimal2CharStr(String hexStr) {
        int len = hexStr.length();
        char[] charArr = new char[len/2];
        for (int i=0; i < len - 1; i += 2) {
            // e.g "41" [as 0x41] -> 4 * 2^4 + 1 = int(65) -> (char)65 = 'A'
            charArr[i/2] = (char)(
                (Character.digit(hexStr.charAt(i), 16) << 4)
                    + Character.digit(hexStr.charAt(i + 1), 16));
        }
        return String.valueOf(charArr);
    }

    // another implementation
    public String process2(String msg1, String msg2, String msg3) {
        String tmp = xor(msg1, msg2);
        return decode(xor(tmp, msg3));
    }

    private String xor(String hexStr1, String hexStr2) {
        StringBuilder builder = new StringBuilder();
        for(int i = 0; i < hexStr1.length(); i += 2) {
            int nA = Integer.parseInt(hexStr1.substring(i, i+2), 16);
            int nB = Integer.parseInt(hexStr2.substring(i, i+2), 16);
            String hex = Integer.toHexString(nA ^ nB);
            // ensure 2 digits via padding
            hex = hex.length() == 1 ? "0" + hex : hex;
            builder.append(hex);
        }
        return builder.toString();
    }

    private String decode(String a) {
        StringBuilder out = new StringBuilder();
        for(int i = 0; i < a.length(); i+=2) {
            // "FF" -> 255
            int n = Integer.parseInt(a.substring(i, i+2), 16);
            // ASCII 65 -> "A"
            out.append((char)n);
        }
        return out.toString();
    }
}
// spend 2 hrs

File: codingame/python/HackingRobberCity.py
import unittest

def hacking_pair_encode(msg1: str, msg2: str, msg3: str) -> str:
    res = []
    # hex string to int each of which repsents a byte
    for i in range(0, len(msg1), 2):
        b1 = int(msg1[i:i+2], 16)
        b2 = int(msg2[i:i+2], 16)
        b3 = int(msg3[i:i+2], 16)
        b = b1 ^ b2 ^ b3
        res.append(chr(b))
    return ''.join(res)

class TestHackingRobberCity(unittest.TestCase):
    # each test case must start with 'test_'
    #
    def test_simple_short(self):
        msg1 = "391813c092a2d5ac9acb705dfe41be3df08de67d1145cbcc3f"
        msg2 = "03adeae2c8c2f2336c8a8d312733c2456e76e0b2d9068adc3f"
        msg3 = "72d0954e354045f09461dc4c911d0b58ff8963efb12c34303f"
        expected = "Hello bob ! How are you ?"
        # assert
        self.assertEqual(expected, hacking_pair_encode(msg1, msg2, msg3))

    def test_long_inner_bytes(self):
        msg1 = "2f03f64ddb650a4517b689a75c1d149dc38a43f0cf4f332871c1144e185552ccab3b13f6f0e831022f7963f5cf59224f4af60bde421be2be107146967d9c6c86b7ec16f81df38febb0810add94abd31e1e056042de83c790eb4102ed7ffd2419d75b1a90ea84f0af46ea9ae2a3e2b9ec6f0d1b99a0f5200e8f0d7d902f13b96aa50c1a1fc03fe245cc835756b03436396230"
        msg2 = "67c7034ed1cb194b6c4db54783d0ac512dfb21ebeaf87e0c4db88bacd4e9c3627e8ff5c2d462d5bef2f90417586d7aa2d77248fdb37239042905b3ce197a629e013e693d25d7d8a6de673aa244bd1263ac93a7998267770fb0366a05c186e2d50be5f04d190211c0d1982f45cf890063d2f326e8d37edb8dc6fabc87fba324259d91610ddae0ff7162991e0e603436396230"
        msg3 = "0a90a22f2ada7c2e08925190b3a4deb5ce1e176905d13850490bfac2a9c4f2c6b4da815157a6c4d0b8f44091b7412b88bdc50670d11eb2ce5154982144d73e2a82ff1dac4c04273f0790510bb536aa18cbb6f2bf3ed788fb6e14588e8f4da3afec8b8eb996e3d758f24681915d5cd8bf84ca584717efcbe171c4f671b185f9290aa9497729ba7c0c9d282c6be03436396230"
        expected = "BTW, to simplify our future exchanges, let's use AES with my 1024-bit private key 5db38d5c0f16ec05ddee67e44617a094e6dd0b837fe5df242e3ea832e30469b0"
        # assert
        self.assertEqual(expected, hacking_pair_encode(msg1, msg2, msg3))

if __name__ == "__main__":
    unittest.main()